openqaoa/optimizers/optimization_methods.py

```diff
@@ -428,3 +428,132 @@ def stochastic_grad_descent(fun, x0, jac, args=(), stepsize=0.001, mass=0.9, sta
 
     return OptimizeResult(x=x, fun=fun(x), jac=g, nit=i+1, nfev=i+1, success=True)
 '''
+
+
+
+def CANS(fun, x0, args=(), maxfev=None, stepsize=0.00001, n_shots_min=10, n_shots_max=10000, mu=0.99, b=1e-06,
+                 maxiter=100, tol=10**(-6), jac=None, jac_w_variance=None,callback=None, **options): #jac_w_uncertainty=None
+
+    print('hi')
+
+    chi = np.zeros(len(x0))
+    xi = 0
+    n_shots = n_shots_min
+
+    bestx = x0
+    besty = np.real(fun(bestx, *args))
+    funcalls = 1  # tracks no. of function evals.
+    niter = 0
+    improved = True
+    stop = False
+
+    testx = np.copy(bestx)
+    testy = besty
+    while improved and not stop and niter < maxiter:
+        lipschitz = np.sum(np.abs(testx)) #TODO there is something wrong with this const
+
+        if not stepsize < 2/lipschitz:
+            print("Stepsize is bigger than 2/Lipschitz: it should be smaller than {0:.3g}".format(2/lipschitz))
+            break
+        
+        print('a',n_shots,testy)
+
+        # compute gradient and variance
+        gradient, variance = jac_w_variance(testx, n_shots=n_shots)
+
+        # compute gradient descent step
+        testx = testx - stepsize*gradient
+        testy = np.real(fun(testx, *args))
+
+        # compute n_shots next step
+        chi = mu*chi + (1-mu)*gradient
+        xi = mu*xi + (1-mu)*np.sum(variance)
+        n_shots = int(np.ceil(2*lipschitz*stepsize*xi/((2-lipschitz*stepsize)*(np.linalg.norm(chi)**2+b*mu**niter))))
+        n_shots = max(n_shots, n_shots_min)
+        n_shots = min(n_shots, n_shots_max) if n_shots_max else n_shots
```

Comment:
Should I save the number of shots at each step in the Logger? Otherwise the user can't know how many shots have been used